src/main.rs

```diff
@@ -131,131 +133,55 @@ fn metadata(path: &Path) -> Result<Metadata, Error> {
-                        .required(true),
-                )
-                .arg(
-                    Arg::with_name("path")
-                        .index(1)
-                        .value_name("path")
-                        .help("Path to check"),
-                ),
-        )
-        .get_matches();
+    let args = cli::parse();
 
-    if let Some(matches) = matches.subcommand_matches("sweep") {
-        let verbose = matches.is_present("verbose");
-        setup_logging(verbose);
+    let criterion = args.criterion();
+    let dry_run = args.dry_run;
+    setup_logging(args.verbose);
 
-        let dry_run = matches.is_present("dry-run");
+    let path = args
+        .path
+        .unwrap_or_else(|| env::current_dir().expect("Failed to get current directory"));
 
-        // Default to current invocation path.
-        let path = match matches.value_of("path") {
-            Some(p) => PathBuf::from(p),
-            None => env::current_dir().expect("Failed to get current directory"),
-        };
+    if let Criterion::Stamp = criterion {
+        debug!("Writing timestamp file in: {:?}", path);
+        return Timestamp::new()
+            .store(path.as_path())
+            .context("Failed to write timestamp file");
+    }
 
-        if matches.is_present("stamp") {
-            debug!("Writing timestamp file in: {:?}", path);
-            return Timestamp::new()
-                .store(path.as_path())
-                .context("Failed to write timestamp file");
+    let paths = if args.recursive {
+        find_cargo_projects(&path, args.hidden)
+    } else {
+        let metadata = metadata(&path).context(format!(
+            "Failed to gather metadata for {:?}",
+            path.display()
+        ))?;
+        let out = Path::new(&metadata.target_directory);
+        if out.exists() {
+            vec![out.into()]
+        } else {
+            anyhow::bail!("Failed to clean {:?} as it does not exist.", out);
         }
+    };
 
-        let paths = if matches.is_present("recursive") {
-            find_cargo_projects(&path, matches.is_present("hidden"))
-        } else {
-            let metadata = metadata(&path).context(format!(
-                "Failed to gather metadata for {:?}",
-                path.display()
-            ))?;
-            let out = Path::new(&metadata.target_directory).to_path_buf();
-            if out.exists() {
-                vec![out]
-            } else {
-                anyhow::bail!("Failed to clean {:?} as it does not exist.", out);
-            }
-        };
+    match criterion {
+        Criterion::Stamp => unreachable!(),
+        Criterion::Installed | Criterion::Toolchains(_) => {
+            let toolchains = match criterion {
+                Criterion::Toolchains(vec) => Some(vec),
+                Criterion::Installed => None,
+                _ => unreachable!(),
+            };
+
+            let hashed_toolchains = hash_toolchains(toolchains.as_ref())?;
```

Comment:
This is what's causing the error - it's now emitting the error to stderr instead of stdout.

Reply:
(I think that's a good change, you just need to update the test assertion.)

Reply:
That's for spotting this, the thing I'm most confused about is `error_output`, now the `run` command is panicking, previously it was not, it's failing before the assertion because of this panic.